src/test/java/itest/RulesPostFormIT.java

```diff
@@ -0,0 +1,149 @@
+ *
+ * (a) the Software, and
+ * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
+ * one is included with the Software (each a "Larger Work" to which the Software
+ * is contributed by such licensors),
+ *
+ * without restriction, including without limitation the rights to copy, create
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package itest;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import io.cryostat.net.web.http.HttpMimeType;
+
+import io.vertx.core.MultiMap;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.core.json.JsonObject;
+import itest.bases.StandardSelfTest;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+class RulesPostFormIT extends StandardSelfTest {
+
+    static MultiMap testRule;
+
+    @BeforeAll
+    static void setup() throws Exception {
+        testRule = MultiMap.caseInsensitiveMultiMap();
+        testRule.add("name", "Test Rule");
+        testRule.add("targetAlias", "es.andrewazor.demo.Main");
+        testRule.add("description", "AutoRulesIT automated rule");
+        testRule.add("eventSpecifier", "template=Continuous,type=TARGET");
+    }
+
+    @Test
+    void testAddRuleThrowsWhenFormAttributesNull() throws Exception {
+        CompletableFuture<JsonObject> response = new CompletableFuture<>();
+
+        webClient
+                .post("/api/v2/rules")
+                .putHeader(HttpHeaders.CONTENT_TYPE.toString(), HttpMimeType.MULTIPART_FORM.mime())
+                .sendForm(
+                        null,
+                        ar -> {
+                            if (assertRequestStatus(ar, response)) {
+                                MatcherAssert.assertThat(
+                                        ar.result().statusCode(), Matchers.equalTo(400));
+                            }
+                        });
```

Comment:
There should be a `response.get()` at the end here to ensure that the test method waits for the request execution to complete before proceeding to the next test.

Reply:
IMO it's better to have the request handler complete the future and perform the assertion on the future value at the end. Here's a snippet from the `AutoRulesIT` showing that pattern:

```java
        CompletableFuture<JsonObject> preRules = new CompletableFuture<>();
        webClient
                .get("/api/v2/rules")
                .send(
                        ar -> {
                            if (assertRequestStatus(ar, preRules)) {
                                preRules.complete(ar.result().bodyAsJsonObject());
                            }
                        });
        JsonObject expectedPreRules =
                new JsonObject(
                        Map.of(
                                "meta", Map.of("type", HttpMimeType.JSON.mime(), "status", "OK"),
                                "data", Map.of("result", List.of())));
        MatcherAssert.assertThat(preRules.get(), Matchers.equalTo(expectedPreRules));
```

Reason being that this pattern ensures that the request is fired and completes, either successfully or not, before the test executor moves on to the next method or proceeds any further down the method. If the assertion is performed asynchronously within the response handler, then any failure mode where a response is simply never received will implicitly pass because the assertion is never executed either.

Reply:
Does that mean I also need `response.get()` after every request, including ones that return an error code?

Reply:
Sorry, didn't see your last comment before I wrote the previous one. Makes sense, I'll get started on refactoring